### ISSUE TITLE
A Static JS file can never receive a 304 Not Modified

### ISSUE THREAD

**Issue by alexkemp9** [openstreetmap/operations#309]
(***Note before:*** This is the inverse of the [previous problem](https://github.com/openstreetmap/openstreetmap-website/issues/2262))    
(The other JS files had cache+expire, no eTag nor LM)    
(This JS file has NO cache+expire, and full eTag+LM)    

*Background*    
Conducting a simple refresh (f5) on a recently downloaded [test map page](https://www.openstreetmap.org/way/17236956) causes only half (10 out of 21) of the files to give a `304 Not Modified`.

Some of those files may well be dynamic, but some are clearly static. Because of the wide range of *Response Header* symptoms + my ignorance these are individual filetype reports to allow those with a working knowledge of the code to decipher it all. Please also note that since these tests the [`Last-Modified` value](https://github.com/openstreetmap/operations/issues/305) has been re-enabled & `eTag` disabled.

*The meat*    
(probably) a static JS file.    

1 x JS    
eg https://piwik.openstreetmap.org/piwik.js

- `cache-control`: none
- `eTag`: strong    
- `Expires`: none
- `Last-Modified`: yes    
- Request: (If-None-Match + If-Modified-Since are both present, but no cache, so no 304)    

*Related*    
The tiles supplied by *toothless* (eg https://b.tile.openstreetmap.org/18/130245/85447.png) are all dynamic yet all have a strong `eTag` (matched by the `If-None-Match`) + `Expires` and thus get cached + a 304.

*Extra*    
See bottom of [this Diary page](https://www.openstreetmap.org/user/alexkemp/diary/368814) for fuller details.

**Comment by tomhughes** [comment]
That file is a static asset served by apache with a trivially simple config so I see no reason why it should behave unusually - indeed my browser seems to be happy to cache it.

Unless there is some magic non-default apache option that needs enabling to get the behaviour you desire I don't believe there is anything we can do here.

**Comment by tomhughes** [comment]
Can you please give a concrete example of something you believe to be wrong - so a URL and set of headers that you sent, a response you got, and the response you think you should have got?

**Comment by tomhughes** [comment]
If I'm understanding the issue correctly then this looks to be a bug/feature in apache related to the interaction between `mod_deflate` and etags. If I request that asset without allowing compression I get:

```
HTTP/1.1 200 OK
Date: Mon, 17 Jun 2019 14:41:16 GMT
Server: Apache/2.4.29 (Ubuntu)
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
Expect-CT: max-age=0, report-uri="https://openstreetmap.report-uri.com/r/d/ct/reportOnly"
Upgrade: h2
Connection: Upgrade
Last-Modified: Tue, 19 Mar 2019 04:53:12 GMT
ETag: "106ad-5846b483ede00"
Accept-Ranges: bytes
Content-Length: 67245
Vary: Accept-Encoding
Content-Type: application/javascript
```

and if I try again with `If-None-Match: "106ad-5846b483ede00"` then I get a 304 as expected:

```
HTTP/1.1 304 Not Modified
Date: Mon, 17 Jun 2019 14:41:58 GMT
Server: Apache/2.4.29 (Ubuntu)
Connection: Upgrade
ETag: "106ad-5846b483ede00"
```

but if I do a request accepting gzip then I get a gzipped body and a different etag:

```
HTTP/1.1 200 OK
Date: Mon, 17 Jun 2019 14:42:43 GMT
Server: Apache/2.4.29 (Ubuntu)
Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
Expect-CT: max-age=0, report-uri="https://openstreetmap.report-uri.com/r/d/ct/reportOnly"
Upgrade: h2
Connection: Upgrade
Last-Modified: Tue, 19 Mar 2019 04:53:12 GMT
ETag: "106ad-5846b483ede00-gzip"
Accept-Ranges: bytes
Vary: Accept-Encoding
Content-Encoding: gzip
Content-Length: 22700
Content-Type: application/javascript
```

and requesting with that etag (even with gzip allowed) doesn't give a 304 response, while asking with the normal etag (without the `-gzip` suffix) does give a 304 response!

**Comment by tomhughes** [comment]
Per https://httpd.apache.org/docs/trunk/mod/mod_deflate.html#deflatealteretag this is apparently apache's default behaviour - you can change it back to not adding the extension but it technically violates the HTTP spec.

**Comment by alexkemp9** [comment]
This is the file [https://www.openstreetmap.org/assets/application-….js](https://www.openstreetmap.org/assets/application-7dc66b028055a8a9bfe43cf7d19b6a06b3a64a54efcbb1ae996f40ae3d5209c1.js) requested using the Request header `Pragma: no-cache` to ensure a *200 OK* response (notice the GZIP):

    HTTP/2.0 200 OK
    date: Mon, 17 Jun 2019 15:41:37 GMT
    server: Apache/2.4.29 (Ubuntu)
    strict-transport-security: max-age=31536000; includeSubDomains; preload
    expect-ct: max-age=0, report-uri="https://openstreetmap.report-uri.com/r/d/ct/reportOnly"
    last-modified: Thu, 30 May 2019 09:00:14 GMT
    accept-ranges: bytes
    cache-control: max-age=31536000
    expires: Tue, 16 Jun 2020 15:41:37 GMT
    vary: Accept-Encoding
    content-encoding: gzip
    content-type: application/javascript
    X-Firefox-Spdy: h2

That file again after an ordinary request; first the relevant Request Headers:

    Accept-Encoding: gzip, deflate, br
    If-Modified-Since: Thu, 30 May 2019 09:00:14 GMT

…then Response:

    HTTP/2.0 304 Not Modified
    date: Mon, 17 Jun 2019 15:51:40 GMT
    server: Apache/2.4.29 (Ubuntu)
    expires: Tue, 16 Jun 2020 15:51:40 GMT
    cache-control: max-age=31536000
    X-Firefox-Spdy: h2

Neither gzipped `eTag` nor `last-modified` have anything to do with it (check your reported `eTag` with/without the '-gzip' & it is identical, since this is a strong `eTag`). The problem is the lack of a *Cache* or *Expires*. This file is NOT cached and, therefore, it cannot be restored from cache. That is hardly difficult.

You need to implement a `cache-control` that is not 'none' + an `expires` that is sane. If you then find that it will not 304 with the `eTag` suffix (definitely a bug, since this is a transport encoding & not a storage encoding) then either switch off the `eTag` or remove the 'gzip' addition.

I do not understand what is difficult about this.

**Comment by alexkemp9** [comment]
Just to say:    

One of the points of introducing `eTag` was situations where content would change but `last-modified` remained the same. Unless you are proposing to do that with this .js file I do not foresee a problem in dropping `eTag`.

**Comment by tomhughes** [comment]
This ticket is about `piwik.js` - please don't confuse it by discussing other assets which are served from completely different servers using completely different rules.

The main problem with the main web site assets is actually the opposite - there are three different servers each of which has slightly different last modified times so If-Modified-Since requests often fail because they hit a different server.

Right now you will find they will work because I have hacked the servers to all have the same times bit I am still considering what we can do as a long term fix.

**Comment by alexkemp9** [comment]
I do not think I ever remembered to say it to Tom:

Making the `Last-Modified` change completely fixed all problems I reported yesterday with all files except HTML/XML + this one. Many, many thanks for a quick action. I was so close to each problem I never suspected a single change could fix every one.

Thanks again.

**Comment by tomhughes** [comment]
The web site assets aren't deployed by rsync - each server generates the published assets independently as part of it's deployment cycle. They also have expires, but that apparently doesn't stop browsers wanting to revalidate.

The piwik script doesn't have expires because it doesn't have a unique name and we don't know when it might change so it has to use a more normal revalidation system, it's just that the deflate bug is defeating that. I have no idea why it can't match both etags!

As I say I have some ideas for both issues but I'm still experimenting...

**Comment by alexkemp9** [comment]
Hmm, just looked at the [Apache-2.5 config doc for etag-gzip](https://httpd.apache.org/docs/trunk/mod/mod_deflate.html#deflatealteretag). I see the point about no-304s clearly flagged there (would have been a good idea for me to read that before running off at the mouth, I guess). That is a bad bug, y'know.

 >   ***AddSuffix***
>    Append the compression method onto the end of the ETag, causing compressed and uncompressed representations to have unique ETags. This has been the default since 2.4.0, but prevents serving "HTTP Not Modified" (304) responses to conditional requests for compressed content.

> ***NoChange***
> Don't change the ETag on a compressed response. This was the default prior to 2.4.0, but does not satisfy the HTTP/1.1 property that all representations of the same resource have unique ETags.

> ***Remove***
> Remove the ETag header from compressed responses. This prevents some conditional requests from being possible, but avoids the shortcomings of the preceding options.

Rather than using the current *AddSuffix* I would suggest either *NoChange* or *Remove*. 

The *transport-encoding* used makes zero difference to the end product at the browser, so why change the `eTag`? The one problem I can think of is with proxy servers that cache the file gzipped & are too brain-dead to take account of `vary: Accept-Encoding`; do they exist?

Also, grating as it is to return to HTTP/1.0 at least that works well, and this is the sole server, so that is another option.

I'm interested to hear your final solution (whoops, heard that somewhere else I think).

**Comment by alexkemp9** [comment]
[RFC7232 (2014)](https://tools.ietf.org/html/rfc7232) seems to be the current RFC for http/1.1 Conditional Requests:

> The Hypertext Transfer Protocol (HTTP) is a stateless application-level protocol for distributed, collaborative, hypertext information systems.  This document defines HTTP/1.1 conditional requests, including metadata header fields for indicating state changes, request header fields for making preconditions on such state, and rules for constructing the responses to a conditional request when one or more preconditions evaluate to false.

…and [RFC7232 sec2.1](https://tools.ietf.org/html/rfc7232#section-2.1) refers directly to weak v strong validators.

**Comment by tomhughes** [comment]
Permanent fixes deployed:

* https://github.com/openstreetmap/chef/commit/8490395fe10856053dba05774cda6c2ffe52aa6e
* https://github.com/openstreetmap/chef/commit/b73a91d8bc8d8f4136753a8f595e419d96eb21b7
* https://github.com/openstreetmap/chef/commit/8adb918189ec57bfe0ef75190e35b75ba6895aae

**Comment by alexkemp9** [comment]
Feedback:    
Now supplying 304 on simple refresh. Thank you!

Origin Server supplies only (strong) etag (former Last-Modified supply is now dropped), but now also supplies `Cache-Control: max-age=604800` + `Expires: Tue, 25 Jun 2019 14:30:35 GMT` and end result is a 304 after refresh.

I do not know whether the next info is connected, but all other fixes reported by me on the same day & fixed by the separate [Last-Modified addition](https://github.com/openstreetmap/operations/issues/305) have now reverted to no-304. Sigh.

**Comment by tomhughes** [comment]
Yes I removed the last modified because it's not going to be reliable as I explained, and switched to using an etag instead.

**Comment by tomhughes** [comment]
Oh that cache control and expires are for piwik (I do wish you'd be clear which resources you are talking about) and that I think sends both last modified and etag. The expiry is only a week because that resource is not fingerprinted and might change occasionally.

**Comment by alexkemp9** [comment]
Sorry to trouble you unnecessarily, Tom.

I noticed that those static files have `cache-control: immutable` included which gives a greyed-out 200 (to mean: not actually re-requested). I was reading up on that (not met it before) and was just navigating back to this page to change my previous comment when you posted.

I did not specify which file the cache & expires referred to as this issue is only about the piwik js file.
